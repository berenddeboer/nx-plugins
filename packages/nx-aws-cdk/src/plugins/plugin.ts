import {
  CreateNodesContextV2,
  CreateNodesV2,
  TargetConfiguration,
  createNodesFromFiles,
} from "@nx/devkit"
import { readdirSync, existsSync } from "fs"
import { dirname, join } from "path"

// Expected format of the plugin options defined in nx.json
export interface CdkPluginOptions {
  synthTargetName?: string
  deployTargetName?: string
  diffTargetName?: string
  rollbackTargetName?: string
  watchTargetName?: string
  destroyTargetName?: string
}

// File glob to find all the configuration files for this plugin
const cdkConfigGlob = "**/cdk.json"

// Entry function that Nx calls to modify the graph
export const createNodesV2: CreateNodesV2<CdkPluginOptions> = [
  cdkConfigGlob,
  async (configFiles, options, context) => {
    return await createNodesFromFiles(
      (configFile, options, context) => createNodesInternal(configFile, options, context),
      configFiles,
      options,
      context
    )
  },
]

async function createNodesInternal(
  configFilePath: string,
  options: CdkPluginOptions,
  context: CreateNodesContextV2
) {
  const projectRoot = dirname(configFilePath)
  const fullPath = join(context.workspaceRoot, projectRoot)

  // Bail early if directory doesn't exist
  if (!existsSync(fullPath)) {
    return {}
  }

  // Do not create a project if package.json or project.json isn't there.
  const siblingFiles = readdirSync(fullPath)
  if (!siblingFiles.includes("package.json") && !siblingFiles.includes("project.json")) {
    return {}
  }

  // Inferred task final output
  const synthTarget: TargetConfiguration = {
    command: `npx cdk synth`,
    options: { cwd: projectRoot },
    cache: true,
    inputs: [
      "default",
      {
        externalDependencies: ["aws-cdk"],
      },
    ],
    outputs: [`{projectRoot}/dist/${projectRoot}`],
  }
  const deployTarget: TargetConfiguration = {
    command: `npx cdk deploy`,
    options: { cwd: projectRoot },
    inputs: [
      "default",
      {
        externalDependencies: ["aws-cdk"],
      },
    ],
    outputs: [`{projectRoot}/dist/${projectRoot}`],
  }
  const diffTarget: TargetConfiguration = {
    ...deployTarget,
    command: `npx cdk diff`,
  }
  const rollbackTarget: TargetConfiguration = {
    ...deployTarget,
    command: `npx cdk rollback`,
  }
  const watchTarget: TargetConfiguration = {
    ...deployTarget,
    command: `npx cdk watch`,
  }
  const destroyTarget: TargetConfiguration = {
    ...deployTarget,
    command: `npx cdk destroy`,
  }

  // Project configuration to be merged into the rest of the Nx configuration
  return {
    projects: {
      [projectRoot]: {
        targets: {
          [options.synthTargetName]: synthTarget,
          [options.deployTargetName]: deployTarget,
          [options.diffTargetName]: diffTarget,
          [options.rollbackTargetName]: rollbackTarget,
          [options.watchTargetName]: watchTarget,
          [options.destroyTargetName]: destroyTarget,
        },
      },
    },
  }
}
