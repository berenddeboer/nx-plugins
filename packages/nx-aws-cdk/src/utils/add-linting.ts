import { Tree } from "nx/src/generators/tree"
import { Linter, LinterType, lintProjectGenerator } from "@nx/eslint"
import { joinPathFragments } from "nx/src/utils/path"
import { GeneratorCallback, runTasksInSerial } from "@nx/devkit"

// TODO(colum): Look into the recommended set up using `withNuxt` inside eslint.config.mjs. https://eslint.nuxt.com/packages/config
export async function addLinting(
  host: Tree,
  options: {
    linter: Linter | LinterType
    projectName: string
    projectRoot: string
    unitTestRunner?: "jest" | "vitest" | "none"
    rootProject?: boolean
  }
) {
  const tasks: GeneratorCallback[] = []
  if (options.linter === "eslint") {
    const lintTask = await lintProjectGenerator(host, {
      linter: options.linter,
      project: options.projectName,
      tsConfigPaths: [joinPathFragments(options.projectRoot, "tsconfig.json")],
      unitTestRunner: options.unitTestRunner,
      skipFormat: true,
      rootProject: options.rootProject,
      addPlugin: true,
    })
    tasks.push(lintTask)
  }
  return runTasksInSerial(...tasks)
}
