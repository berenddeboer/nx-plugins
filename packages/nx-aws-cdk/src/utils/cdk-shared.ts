export const CDK_VERSION = "^2.79.1"
export const CONSTRUCTS_VERSION = "^10.0.0"
export const CDK_ESLINT_VERSION = "^1.8.0"
