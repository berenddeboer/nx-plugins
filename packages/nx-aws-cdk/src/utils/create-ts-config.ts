import { Tree, writeJson } from "@nx/devkit"
import * as shared from "@nx/js/src/utils/typescript/create-ts-config"

// Define interface for TypeScript configuration
interface TsConfig {
  files?: string[]
  references?: Array<{ path: string }>
  compileOnSave?: boolean
  compilerOptions?: Record<string, unknown>
  exclude?: string[]
  extends?: string
  include?: string[]
}

export function createTsConfig(
  host: Tree,
  options: {
    projectRoot: string
    rootProject?: boolean
    unitTestRunner?: string
    isUsingTsSolutionConfig: boolean
  },
  relativePathToRootTsConfig: string
) {
  createAppTsConfig(host, options)
  const json: TsConfig = {
    files: [],
    references: [
      {
        path: "./tsconfig.app.json",
      },
    ],
  }

  if (options.unitTestRunner !== "none") {
    json.references?.push({
      path: "./tsconfig.spec.json",
    })
  }

  // inline tsconfig.base.json into the project
  if (options.rootProject) {
    json.compileOnSave = false
    json.compilerOptions = {
      ...shared.tsConfigBaseOptions,
      ...json.compilerOptions,
    }
    json.exclude = ["node_modules", "tmp", "cdk.out"]
  } else {
    json.extends = relativePathToRootTsConfig
  }

  writeJson(host, `${options.projectRoot}/tsconfig.json`, json)
}

function createAppTsConfig(host: Tree, options: { projectRoot: string }) {
  const json: TsConfig = {
    extends: "./tsconfig.json",
    compilerOptions: {
      composite: true,
    },
    include: ["src/**/*"],
    exclude: [],
  }

  writeJson(host, `${options.projectRoot}/tsconfig.app.json`, json)
}
