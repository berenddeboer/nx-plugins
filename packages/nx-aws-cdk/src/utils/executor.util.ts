import { exec } from "child_process"
import { readFileSync } from "fs"

import { CdkExecutorSchema } from "../executors/cdk/schema"
import { ParsedExecutorInterface } from "../interfaces/parsed-executor.interface"
import { logger, detectPackageManager } from "@nx/devkit"

export const executorPropKeys = ["stacks"]
export const LARGE_BUFFER = 1024 * 1000000

export function parseArgs(options: CdkExecutorSchema): Record<string, string> {
  const keys = Object.keys(options)
  return keys
    .filter((prop) => executorPropKeys.indexOf(prop) < 0)
    .reduce((acc, key) => ((acc[key] = options[key]), acc), {})
}

export function createCommand(command: string, options: ParsedExecutorInterface): string {
  // We cannot start cdk in a monorepo, else it won't detect imported packages
  // Workaround provided here: https://github.com/adrian-goe/nx-aws-cdk-v2/issues/679
  // This improves upon that by detecting the main file to pass to CDK
  // from cdk.json.
  const NX_WORKSPACE_ROOT = process.env.NX_WORKSPACE_ROOT ?? ""
  const packageManager = detectPackageManager()
  const packageManagerExecutor =
    packageManager === "npm" ? "npx" : `${packageManager} dlx`
  const projectPath = `${NX_WORKSPACE_ROOT}/${options.root}`
  const cdk_json = JSON.parse(readFileSync(`${projectPath}/cdk.json`).toString())
  const app: string = cdk_json["app"]
  const main = app.split(" ").pop()
  const generatePath = `"${packageManagerExecutor} ts-node --require tsconfig-paths/register --project ${projectPath}/tsconfig.app.json ${projectPath}/${main}"`
  const cdk = `node --require ts-node/register ${NX_WORKSPACE_ROOT}/node_modules/aws-cdk/bin/cdk.js -a ${generatePath}`
  let commands = [cdk, command]

  if (options.stacks) {
    commands = commands.concat(options.stacks)
  }
  if (options.context) {
    Object.keys(options.context).forEach((key) => {
      commands.push(`--context ${key}=${options.context[key]}`)
    })
  }

  for (const arg in options.otherArgs) {
    commands.push(`--${arg}=${options.otherArgs[arg]}`)
  }

  return commands.join(" ")
}

export function runCommandProcess(command: string, cwd: string): Promise<boolean> {
  return new Promise((resolve) => {
    logger.debug(`Executing command: ${command}`)

    const childProcess = exec(command, {
      maxBuffer: LARGE_BUFFER,
      env: process.env,
      cwd: cwd,
    })

    // Ensure the child process is killed when the parent exits
    const processExitListener = () => childProcess.kill()
    process.on("exit", processExitListener)
    process.on("SIGTERM", processExitListener)

    process.stdin.on("data", (data) => {
      childProcess.stdin.write(data)
      childProcess.stdin.end()
    })

    childProcess.stdout.on("data", (data) => {
      process.stdout.write(data)
    })

    childProcess.stderr.on("data", (err) => {
      process.stderr.write(err)
    })

    childProcess.on("close", (code) => {
      if (code === 0) {
        resolve(true)
      } else {
        resolve(false)
      }

      process.removeListener("exit", processExitListener)
      process.removeListener("SIGTERM", processExitListener)

      process.stdin.removeListener("data", processExitListener)
    })
  })
}
