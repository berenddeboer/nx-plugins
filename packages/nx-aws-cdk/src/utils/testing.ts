import { ExecutorContext } from "@nx/devkit"

export function mockExecutorContext(
  executorName: string,
  workspaceVersion = 2
): ExecutorContext {
  return {
    projectName: "proj",
    root: "/root",
    cwd: "/root",
    workspace: {
      version: workspaceVersion,
      projects: {
        proj: {
          root: "apps/proj",
          targets: {
            test: {
              executor: `@berenddeboer/nx-aws-cdk:${executorName}`,
            },
          },
        },
      },
    },
    target: {
      executor: `@berenddeboer/nx-aws-cdk:${executorName}`,
    },
    isVerbose: true,
  }
}
