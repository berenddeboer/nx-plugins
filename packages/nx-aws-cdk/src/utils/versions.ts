// eslint-disable-next-line @typescript-eslint/no-var-requires
export const nxVersion = require("../../package.json").version

export const CDK_VERSION = "^2.179.0"
export const CDK_CLI_VERSION = "^2.179.0"
export const CONSTRUCTS_VERSION = "^10.3.0"
export const CDK_ESLINT_VERSION = "^1.8.0"
export const TS_NODE_VERSION = "^10.9.2"
