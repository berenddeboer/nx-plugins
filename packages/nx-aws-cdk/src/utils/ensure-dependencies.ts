import type { GeneratorCallback, Tree } from "@nx/devkit"
import { CDK_CLI_VERSION, CDK_VERSION, CONSTRUCTS_VERSION } from "./versions"
import { addDependenciesToPackageJson } from "@nx/devkit"

export function ensureDependencies(tree: Tree): GeneratorCallback {
  return addDependenciesToPackageJson(
    tree,
    {
      "aws-cdk-lib": CDK_VERSION,
      constructs: CONSTRUCTS_VERSION,
    },
    {
      "aws-cdk": CDK_CLI_VERSION,
    }
  )
}
