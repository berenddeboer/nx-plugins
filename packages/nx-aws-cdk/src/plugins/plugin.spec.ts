import { CreateNodesContext } from "@nx/devkit"
import { createNodesV2 } from "./plugin"
import { TempFs } from "../utils/temp-fs"

describe("@berenddeboer/nx-aws-cdk/plugin", () => {
  const createNodesFunction = createNodesV2[1]
  let context: CreateNodesContext

  describe("root project", () => {
    const tempFs = new TempFs("test")
    const projectRoot = "stacks/cdk-app"

    beforeEach(async () => {
      context = {
        nxJsonConfiguration: {
          // These defaults should be overridden by plugin
          targetDefaults: {
            build: {
              cache: false,
              inputs: ["foo", "^foo"],
            },
          },
          namedInputs: {
            default: ["{projectRoot}/**/*"],
            production: ["!{projectRoot}/**/*.spec.ts"],
          },
        },
        workspaceRoot: tempFs.tempDir,
        configFiles: [],
      }

      tempFs.createFileSync(
        `${projectRoot}/package.json`,
        JSON.stringify({ name: "cdk-app" })
      )
    })

    afterEach(() => {
      jest.resetModules()
      tempFs.cleanup()
    })

    it("should create all targets when names are provided", async () => {
      const options = {
        cdkTargetName: "cdk",
        synthTargetName: "synth",
        deployTargetName: "deploy",
        diffTargetName: "diff",
        rollbackTargetName: "rollback",
        watchTargetName: "watch",
        destroyTargetName: "destroy",
      }
      const nodes = await createNodesFunction(
        [`${projectRoot}/cdk.json`],
        options,
        context
      )
      const project = nodes[0][1].projects[projectRoot]

      expect(project.targets).toHaveProperty("cdk")
      expect(project.targets).toHaveProperty("synth")
      expect(project.targets).toHaveProperty("deploy")
      expect(project.targets).toHaveProperty("diff")
      expect(project.targets).toHaveProperty("rollback")
      expect(project.targets).toHaveProperty("watch")
      expect(project.targets).toHaveProperty("destroy")
    })

    it("should not create targets when names are null", async () => {
      const options = {
        synthTargetName: null,
        deployTargetName: null,
        diffTargetName: null,
        rollbackTargetName: null,
        watchTargetName: null,
        destroyTargetName: null,
      }
      const nodes = await createNodesFunction(
        [`${projectRoot}/cdk.json`],
        options,
        context
      )
      const project = nodes[0][1].projects[projectRoot]

      expect(project.targets).toEqual({})
    })

    it("should not create targets when names are empty strings", async () => {
      const options = {
        synthTargetName: "",
        deployTargetName: "",
        diffTargetName: "",
        rollbackTargetName: "",
        watchTargetName: "",
        destroyTargetName: "",
      }
      const nodes = await createNodesFunction(
        [`${projectRoot}/cdk.json`],
        options,
        context
      )
      const project = nodes[0][1].projects[projectRoot]

      expect(project.targets).toEqual({})
    })

    it("should not create targets when names are not specified", async () => {
      const nodes = await createNodesFunction([`${projectRoot}/cdk.json`], {}, context)
      const project = nodes[0][1].projects[projectRoot]

      expect(project.targets).toEqual({})
    })

    it("should create only specified targets", async () => {
      const options = {
        synthTargetName: "synth",
        deployTargetName: "deploy",
        // other targets not specified
      }
      const nodes = await createNodesFunction(
        [`${projectRoot}/cdk.json`],
        options,
        context
      )
      const project = nodes[0][1].projects[projectRoot]

      expect(project.targets).toHaveProperty("synth")
      expect(project.targets).toHaveProperty("deploy")
      expect(project.targets).not.toHaveProperty("diff")
      expect(project.targets).not.toHaveProperty("rollback")
      expect(project.targets).not.toHaveProperty("watch")
      expect(project.targets).not.toHaveProperty("destroy")
    })
  })
})
