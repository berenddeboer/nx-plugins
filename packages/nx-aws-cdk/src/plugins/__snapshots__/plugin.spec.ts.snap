// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@berenddeboer/nx-aws-cdk/plugin root project should create nodes 1`] = `
[
  [
    "cdk.json",
    {
      "projects": {
        ".": {
          "targets": {
            "undefined": {
              "command": "npx cdk destroy",
              "inputs": [
                "default",
                {
                  "externalDependencies": [
                    "aws-cdk",
                  ],
                },
              ],
              "options": {
                "cwd": ".",
              },
              "outputs": [
                "{projectRoot}/dist/.",
              ],
            },
          },
        },
      },
    },
  ],
]
`;
